File: visual-basic/AVL/Arvore.vb
Imports System.Collections
Imports System.Text
Imports System.Runtime.CompilerServices

<Assembly: InternalsVisibleTo("Teste")>

''' <summary>
''' Árvore binária de busca balanceada automaticamente por meio do algoritmo AVL.
''' </summary>
''' <typeparam name="T">Tipo dos elementos armazenados na árvore.</typeparam>
Public Class Arvore(Of T As IComparable)

    Friend Raiz As New No(Of T)(Nothing)
    Public ReadOnly Property Contagem

    ''' <summary>
    ''' Função recursiva que desce na árvore até encontrar o elemento
    ''' comparando ele a cada nó.
    ''' </summary>
    ''' <param name="Elemento">Elemento procurado</param>
    ''' <param name="EsteNo">Nó que será examinado nesta iteração.</param>
    ''' <returns>
    '''     O nó contendo o elemento, se encontrado, 
    '''     ou o nó vazio onde o elemento deveria ser inserido, se não encontrado.
    ''' </returns>
    Private Function Encontrar(Elemento As T, EsteNo As No(Of T)) As No(Of T)

        If (EsteNo.Vazio) Then
            ' Elemento não foi encontrado, mas pode ser adicionado nesta posição.
            Return EsteNo
        End If

        Dim Diferenca = EsteNo.Elemento.CompareTo(Elemento)

        If Diferenca > 0 Then
            ' Elemento vem antes deste nó.
            Return Encontrar(Elemento, EsteNo.Esquerda)
        ElseIf Diferenca < 0 Then
            ' Elemento vem depois deste nó.
            Return Encontrar(Elemento, EsteNo.Direita)
        Else
            ' Elemento encontrado neste nó.
            Return EsteNo
        End If

    End Function

    ''' <summary>
    ''' Verifica se Elemento está presente na árvore.
    ''' </summary>
    ''' <param name="Elemento">Elemento procurado.</param>
    Public Function Contem(Elemento As T) As Boolean
        ' Se o método Encontrar retorna um nó vazio, isso significa que Elemento não foi encontrado.
        Return Not Encontrar(Elemento, Raiz).Vazio
    End Function

    ''' <summary>
    ''' Insere este elemento na árvore. Balanceia a árvore caso necessário.
    ''' </summary>
    ''' <param name="Elemento"></param>
    ''' <returns>True se o elemento foi inserido com sucesso, false se já existe na árvore.</returns>
    Public Function Inserir(Elemento As T) As Boolean
        Dim NoEncontrado = Encontrar(Elemento, Raiz)

        ' Se o nó não é vazio o elemento já existe.
        ' A árvore não aceita elementos duplicados.
        If Not NoEncontrado.Vazio Then Return False

        NoEncontrado.Elemento = Elemento

        _Contagem += 1

        ' Balanceia a subárvore modificada e todos os nós acima.
        If NoEncontrado.Pai IsNot Nothing Then
            Balancear(NoEncontrado.Pai)
        End If

        Return True

    End Function

    ''' <summary>
    ''' Remove o nó da árvore. Balanceia a árvore se necessário.
    ''' </summary>
    ''' <param name="NoARemover"></param>
    Private Sub Remover(NoARemover As No(Of T))

        ' Em 3 dos próximos 4 casos a modificação mais profunda na árvore é no nó removido.
        Dim NoABalancear = NoARemover.Pai

        If NoARemover.Altura = 1 Then
            ' Altura = 1 significa o mesmo que Esquerda.Vazio And Direita.Vazio
            ' Nó folha.
            NoARemover.Esvaziar()

        ElseIf NoARemover.Esquerda.Vazio Then
            ' Tem apenas um filho à direita.
            SubstituirNo(NoARemover, NoARemover.Direita)
            ' Garbage Collector pode liberar o espaço de NoEncontrado.

        ElseIf NoARemover.Direita.Vazio Then
            ' Tem apenas um filho à esquerda.
            SubstituirNo(NoARemover, NoARemover.Esquerda)
            ' Garbage Collector pode liberar o espaço de NoEncontrado.

        Else
            ' Tem dois filhos.

            ' Encontra nó mais à direita na subárvore esquerda.
            Dim Predecessor = NoARemover.Esquerda
            Do Until Predecessor.Direita.Vazio
                Predecessor = Predecessor.Direita
            Loop

            ' Coloca o elemento mais à direita da subárvore esquerda no lugar do elemento removido.
            NoARemover.Elemento = Predecessor.Elemento
            ' A árvore será balanceada quando Predecessor for removido.
            NoABalancear = Nothing
            ' Como o elemento predecessor está no lugar do elemento removido, ele pode ser excluido.
            Remover(Predecessor)
            ' Anulando a modificação dupla da Contagem devido à remoção do Predecessor.
            _Contagem += 1

            ' Garbage Collector pode liberar o espaço de NoEncontrado.

        End If

        _Contagem -= 1

        If (NoABalancear IsNot Nothing) Then
            Balancear(NoABalancear)
        End If

    End Sub

    ''' <summary>
    ''' Remove o elemento da árvore se ele existir. Balanceia a árvore se necessário.
    ''' </summary>
    ''' <param name="Elemento"></param>
    ''' <returns>True se o elemento foi removido, False se não está presente.</returns>
    Public Function Remover(Elemento As T) As Boolean
        Dim NoEncontrado = Encontrar(Elemento, Raiz)

        If NoEncontrado.Vazio Then Return False

        Remover(NoEncontrado)
        Return True

    End Function

    ' Balanceamento

    ''' <summary>
    ''' Verifica o balanço do nó N e, dependendo do valor, realiza rotações para balancear a subárvore.
    ''' </summary>
    ''' <param name="N">Nó que será verificado e balanceado.</param>
    Private Sub Balancear(N As No(Of T))

        N.AtualizarAltura()

        Dim Balanco = N.FatorBalanco

        ' Se -1 <= FatorBalanco <= 1 o nó já está balanceado e nada precisa ser feito.
        If Balanco < -1 Then
            RotacaoADireita(N)
        ElseIf Balanco > 1 Then
            RotacaoAEsquerda(N)
        End If

        If N.Pai IsNot Nothing Then
            Balancear(N.Pai)
        End If

    End Sub

    ' Rotacoes
    ''' <summary>
    ''' O filho à direita se torna a nova raiz da subárvore.
    ''' <code>
    '''       A                 B
    '''   x       B    =>   A       C
    ''' x   x   x   C
    ''' </code>
    ''' </summary>
    ''' <param name="A">Nó que vai ser rotacionado.</param>
    Private Sub RotacaoAEsquerda(A As No(Of T))

        ' Rotação dupla se o balanço de C é menor que 0
        '       A
        '   x       C
        ' x   x   B   x
        If A.Direita.FatorBalanco < 0 Then RotacaoADireita(A.Direita)

        ' Agora que o balanço de B é maior ou igual a 0 fazer uma rotação simples.
        '       A
        '   x       B
        ' x   x   x   C

        Dim B = A.Direita

        ' O filho à esquerda de B, mesmo que seja nó vazio, se torna filho à direita de A.
        A.Direita = B.Esquerda

        ' Se A era raiz da árvore, agora B se torna raiz.
        SubstituirNo(A, B)

        ' A se torna filho à esquerda de B
        B.Esquerda = A

        ' Resultado
        '       B
        '   A       C

        ' Os nós acima desta subárvore serão atualizados mais tarde porque Balancear() chama AtualizarAltura para todos eles.
        A.AtualizarAltura()
        B.AtualizarAltura()
        B.Direita.AtualizarAltura()

    End Sub

    ''' <summary>
    ''' O filho à esquerda se torna a nova raiz da subárvore.
    ''' <code>
    '''       C                 B
    '''   B       x    =>   A       C
    ''' A   x   x   x
    ''' </code>
    ''' </summary>
    ''' <param name="C">Nó que vai ser rotacionado.</param>
    Private Sub RotacaoADireita(C As No(Of T))

        ' Rotação dupla se o balanço de A é maior que 0.
        '       C
        '   A       x
        ' x   B   x   x
        If C.Esquerda.FatorBalanco > 0 Then RotacaoAEsquerda(C.Esquerda)

        ' Agora que o balanço de B é menor ou igual a 0 fazer uma rotação simples.
        '       C
        '   B       x
        ' A   x   x   x

        Dim B = C.Esquerda

        ' O filho à direita de B, mesmo que seja nó vazio, se torna filho à esquerda de C.
        C.Esquerda = B.Direita

        ' Se C era raiz da árvore, agora B se torna raiz.
        SubstituirNo(C, B)

        ' C se torna filho à direita de B.
        B.Direita = C

        ' Resultado
        '       B
        '   A       C

        ' Os nós acima desta subárvore serão atualizados mais tarde porque Balancear() chama AtualizarAltura para todos eles.
        B.Esquerda.AtualizarAltura()
        B.AtualizarAltura()
        C.AtualizarAltura()

    End Sub

    ''' <summary>
    ''' Atualiza o Pai de Novo por colocá-lo no lugar de Antigo e atualiza a raiz da árvore se necessário.
    ''' </summary>
    ''' <param name="Antigo">Nó que deve ser substituido.</param>
    ''' <param name="Novo">Nó que vai ser colocado na posição de antigo.</param>
    Private Sub SubstituirNo(Antigo As No(Of T), Novo As No(Of T))

        ' Se Antigo era a raiz da árvore, agora Novo se torna raiz.
        If Antigo.Pai Is Nothing Then
            Raiz = Novo
            Novo.Pai = Nothing
            ' Se Antigo era um filho à esquerda, Novo o substitui. Compara por referência.
        ElseIf Antigo.Pai.Esquerda Is Antigo Then
            Antigo.Pai.Esquerda = Novo
            ' Se Antigo era filho à direita, Novo o substitui.
        Else
            Antigo.Pai.Direita = Novo
        End If

    End Sub

    ' Métodos auxiliares para Strings.

    ''' <summary>
    ''' Representação da árvore na vertical. Nó direito fica abaixo do esquerdo.
    ''' Ocupa menos espaço que a horizontal.
    ''' </summary>
    Public Function ToStringVertical() As String

        Dim Resultado As New StringBuilder

        ToStringVerticalRecursivo(Resultado, 0, Raiz, "*")

        Return Resultado.ToString

    End Function

    ''' <summary>
    ''' Adiciona cada nó à Representação após a indentação e um indicador de esquerda ou direita.
    ''' </summary>
    ''' <param name="Representacao">StringBuilder que contém a representação da árvore.</param>
    ''' <param name="Nivel">Profundidade do nó que será representada como indentação. 0 para raiz.</param>
    ''' <param name="EsteNo"></param>
    Private Sub ToStringVerticalRecursivo(ByRef Representacao As StringBuilder, Nivel As Integer, EsteNo As No(Of T), Posicao As String)

        If EsteNo.Vazio Then Return

        Representacao.Append(
            StrDup(Nivel, "  ") +
            Posicao + " " +
            EsteNo.Elemento.ToString +
            vbNewLine
        )

        ToStringVerticalRecursivo(Representacao, Nivel + 1, EsteNo.Esquerda, "E")
        ToStringVerticalRecursivo(Representacao, Nivel + 1, EsteNo.Direita, "D")

    End Sub

    ''' <summary>
    ''' Represtentação da árvore em texto na horizontal.
    ''' Nós irmãos ficam lado a lado.
    ''' </summary>
    Public Overrides Function ToString() As String

        If Raiz.Vazio Then Return ""

        ' Para cada Linha da representação é feita uma lista de nós do nível correspondente da árvore
        Dim Linhas As New ArrayList() From {
            New ArrayList From {
                Raiz
            }
        }
        ' A String que será retornada.
        Dim Resultado As New StringBuilder("")
        ' Número de caracteres necessários para representar o maior elemento da árvore.
        Dim LarguraNo = Raiz.Elemento.ToString.Length
        ' Se um dos nós deste nível tem filhos então o loop continua.
        Dim Continuar = True

        While Continuar

            Dim ProximaLinha = New ArrayList
            Continuar = False

            For Each N In Linhas(Linhas.Count - 1)

                If IsNothing(N) Then
                    ProximaLinha.AddRange({Nothing, Nothing})
                Else
                    ' Largura deste nó.
                    Dim NLength = N.Elemento.ToString.Length
                    If NLength > LarguraNo Then
                        LarguraNo = NLength
                    End If

                    ' Prepara a próxima Linha.
                    If N.Esquerda.Vazio Then
                        ProximaLinha.Add(Nothing)
                    Else
                        ProximaLinha.Add(N.Esquerda)
                        Continuar = True
                    End If
                    If N.Direita.Vazio Then
                        ProximaLinha.Add(Nothing)
                    Else
                        ProximaLinha.Add(N.Direita)
                        Continuar = True
                    End If
                End If
            Next

            If Continuar Then
                Linhas.Add(ProximaLinha)
            End If

        End While

        Dim NumNosUltimaLinha = Linhas(Linhas.Count - 1).Count

        ' Largura da última linha em função do espaço de um nó.
        ' Espaçamento na ultima linha é igual a um nó vazio.
        Dim LarguraArvore = (2 * NumNosUltimaLinha) - 1

        Dim EspacoVazio = StrDup(LarguraNo, " ")
        Dim NoVazio = StrDup(LarguraNo, "-")
        Dim Profundidade = Linhas.Count

        For I = 0 To Linhas.Count - 1
            Dim Linha = Linhas(I)

            ' Espaço entre nós em função do espaço de um nó. Ex.: 1 nó vazio, 2 nós vazio, ...
            Dim Espacamento = Math.Pow(2, (Profundidade - I)) - 1


            ' Espaço que é adicionado à esquerda para centralizar o nó nesta linha.
            '       Lagura da árvore em nós.
            '       Número de nós desta linha -
            '       Espaço ocupado nesta linha entre os nós em função do espaço de um nó.
            Dim LarguraDestaLinha = LarguraArvore -
                    Linha.Count -
                    ((Linha.Count - 1) * Espacamento)

            ' Adiciona a indentação antes da linha.
            Resultado.Append(StrDup(
                    LarguraDestaLinha \ 2,
                EspacoVazio
            ))

            For J = 0 To Linha.Count - 1
                Dim N = Linha(J)
                If IsNothing(N) Then
                    Resultado.Append(NoVazio)
                Else
                    Resultado.Append(N.Elemento.ToString.PadLeft(LarguraNo))
                End If
                If J <> Linha.Count - 1 Then
                    Resultado.Append(StrDup(CInt(Espacamento), EspacoVazio))
                End If
            Next
            Resultado.Append(vbNewLine)
        Next

        Return Resultado.ToString


    End Function

End Class

File: java/src/main/java/arvore/Arvore.java
package arvore;

import java.util.ArrayList;
import java.util.List;

/**
 * Árvore binária de busca.
 * Balanceada automaticamente por meio do algoritmo AVL.
 * @param <T> Tipo dos valores armazenados na árvore.
 */
public class Arvore<T extends Comparable<T>> {

    private No<T> raiz = new No<>(null);
    private int contagem = 0;

    // Getters
    public No<T> getRaiz() { return raiz; }

    public int getContagem() { return contagem; }

    /**
     * Função recursiva que desce na árvore até encontrar o valor, comparando ele a cada nó.
     * @param valor Valor procurado.
     * @param no Nó que será examinado nesta iteração.
     * @return O nó contendo o valor, se encontrado,
     * ou o nó vazio onde o valor deveria ser inserido, se não encontrado.
     */
    private No<T> encontrar(T valor, No<T> no) {

        if (no.isVazio()) {
            // valor não foi encontrado, mas pode ser adicionado nesta posição.
            return no;
        }

        int diferenca = no.getValor().compareTo(valor);

        return switch (diferenca) {

            // valor encontrado neste nó.
            case 0 -> no;

            // valor é maior do que este nó.
            case -1 -> encontrar(valor, no.getDireita());

            // Case 1. valor é menor do que este nó.
            default -> encontrar(valor, no.getEsquerda());

        };
    }

    /**
     * Verifica se valor está presente na árvore.
     */
    public boolean contem(T valor) {
        // Se encontrar retorna um nó vazio, isso significa que o valor não foi encontrado.
        return !encontrar(valor, raiz).isVazio();
    }

    /**
     * Insere este valor na árvore. Balanceia caso necessário.
     * @return True se o valor foi inserido com sucesso, false se já existe na árvore.
     */
    public boolean inserir(T valor) {

        No<T> noEncontrado = encontrar(valor, raiz);

        // Se o nó não é vazio, o valor já existe.
        // A árvore não aceita valores duplicados.
        if (!noEncontrado.isVazio()) { return false; }

        noEncontrado.setValor(valor);

        contagem += 1;

        // Balanceia a subárvore modificada e todos os nós acima.
        if (noEncontrado.pai != null) {
            balancear(noEncontrado.pai);
        }

        return true;
    }

    /**
     * Remove o nó da árvore. Balanceia a árvore se necessário.
     */
    private void remover(No<T> no) {

        // Em 3 dos próximos 4 casos a modificação mais profunda na árvore
        // é no nó removido.
        No<T> noABalancear = no.pai;

        if (no.getAltura() == 1) {
            // Altura = 1 significa o mesmo que
            // esquerda.isVazio() && direita.isVazio().
            no.esvaziar();
        } else if (no.getEsquerda().isVazio()) {
            // Tem apenas um filho à direita.
            substituirNo(no, no.getDireita());
            // Garbage Collector pode liberar o espaço de no.
        } else if (no.getDireita().isVazio()) {
            // Tem apenas um filho à esquerda.
            substituirNo(no, no.getEsquerda());
            // Garbage Collector pode liberar o espaço de no.
        } else {
            // Tem dois filhos.

            // Encontra nós mais à direita na subárvore esquerda.
            No<T> predecessor = no.getEsquerda();
            while (!predecessor.getDireita().isVazio()) {
                predecessor =  predecessor.getDireita();
            }

            // Coloca o valor mais alto da subárvore esquerda no lugar do valor removido.
            no.setValor(predecessor.getValor());
            // A árvore será balanceada quando o predecessor for removido.
            noABalancear = null;
            // Como o valor do predecessor está no lugar do valor removido, ele pode ser excluido.
            remover(predecessor);
            // Anulando a modificação dupla da contagem devido à remoção do predecessor.
            contagem += 1;

            // Garbage Collector pode liberar o espaço de no.
        }

        contagem -= 1;

        if (noABalancear != null) {
            balancear(noABalancear);
        }

    }

    /**
     * Remove valor da árvore se ele existir. Balanceia a árvore se necessário.
     * @return True se o valor foi removido, false se não está presente.
     */
    public boolean remover(T valor) {

        No<T> noEncontrado = encontrar(valor, raiz);

        if (noEncontrado.isVazio()) { return false; }

        remover(noEncontrado);
        return true;
    }

    // Balanceamento

    /**
     * Verifica o balanço do nó e, dependendo do valor, realiza rotações para tornar esta subárvore balanceada.
     * @param no Nó que será verificado e balanceado.
     */
    private void balancear(No<T> no) {

        no.atualizarAltura();

        // Substitui a propriedade para evitar recalcular.
        int balanco = no.getFatorBalanco();

        // Se -1 <= fatorBalanco <= 1, o nó já está balanceado e nada precisa ser feito.
        if (balanco < -1) {
            rotacaoADireita(no);
        } else if (balanco > 1) {
            rotacaoAEsquerda(no);
        }

        if (no.pai != null) {
            balancear(no.pai);
        }

    }

    // Rotações

    /**
     * O filho à direita se torna a nova raiz da subárvore.
     * <pre>
     *       a                 b
     *   x       b    =>   a       c
     * x   x   x   c
     * </pre>
     * @param a Nó que vai ser rotacionado.
     */
    private void rotacaoAEsquerda(No<T> a) {

        /* Rotação dupla se o balanço de c é menor do que 0.
         *       a
         *   x       c
         * x   x   b   x
         */
        if (a.getDireita().getFatorBalanco() < 0) { rotacaoADireita(a.getDireita()); }

        /* Agora que o balanço de B é maior ou igual a 0 fazer uma rotação simples.
         *       a
         *   x       b
         * x   x   x   c
         */

        // Salvando referência a b antes de substituií-lo.
        No<T> b = a.getDireita();

        // O filho à esquerda de b, mesmo que seja nó vazio, se torna filho à direita de a.
        a.setDireita(b.getEsquerda());

        // Se a era raiz da árvore, agora b se torna raiz.
        substituirNo(a, b);

        // a se torna filho à esquerda de b.
        b.setEsquerda(a);

        /* Resultado
         *       b
         *   a       c
         */

        // Atualizando as alturas dos nós afetados, a, b e c.
        // Os nós acima desta subárvore serão atualizados mais tarde
        // porque balancear chama atualizarAltura para todos eles.
        a.atualizarAltura();
        b.atualizarAltura();
        b.getDireita().atualizarAltura();
    }

    /**
     * O filho à esquerda se torna a nova raiz da subárvore.
     * <pre>
     *       c                 b
     *   b       x    =>   a       c
     * a   x   x   x
     * </pre>
     * @param c Nó que vai ser rotacionado.
     */
    private void rotacaoADireita(No<T> c) {

        /* Rotação dupla se o balanço de A é maior que 0.
         *       c
         *   a       x
         * x   b   x   x
         */
        if (c.getEsquerda().getFatorBalanco() > 0) { rotacaoAEsquerda(c.getEsquerda()); }

        /* Agora que o balanço de B é menor ou igual a 0 fazer uma rotação simples.
         *       c
         *   b       x
         * a   x   x   x
         */

        // Salvando referência a b antes de substituií-lo.
        No<T> b = c.getEsquerda();

        // O filho à direita de b, mesmo que seja nó vazio, se torna filho à direita de c.
        c.setEsquerda(b.getDireita());

        // Se c era raiz da árvore, agora b se torna raiz.
        substituirNo(c, b);

        // a se torna filho à esquerda de b.
        b.setDireita(c);

        /* Resultado
         *       b
         *   a       c
         */

        // Atualizando as alturas dos nós afetados, a, b e c.
        // Os nós acima desta subárvore serão atualizados mais tarde
        // porque balancear() chama atualizarAltura para todos eles.
        b.atualizarAltura();
        c.atualizarAltura();
        b.getEsquerda().atualizarAltura();
    }

    /**
     * Atualiza o pai de novo por colocá-lo no lugar de antigo e
     * atualiza a raiz da árvore se necessário.
     */
    private void substituirNo(No<T> antigo, No<T> novo) {

        // Se antigo era raiz da árvore, agora novo se torna raiz.
        if (antigo.pai == null) {
            raiz = novo;
            novo.pai = null;
        // Se antigo era um filho à esquerda, novo o substitui.
        } else if (antigo.pai.getEsquerda().equals(antigo)) {
            antigo.pai.setEsquerda(novo);
        // Se antigo era um filho à direita, novo o substitui.
        } else {
            antigo.pai.setDireita(novo);
        }
    }

    // Métodos auxiliares para Strings.

    /**
     * Representação da árvore na vertical. Nó direito fica abaixo do esquerdo.
     * Ocupa menos espaço que a horizontal.
     */
    public String toStringVertical() {

        StringBuilder stringBuilder = new StringBuilder();

        toStringVerticalRecursivo(stringBuilder, 0, raiz, '*');

        return stringBuilder.toString();
    }

    /**
     * Para cada nó, adiciona ao StringBuilder o valor após a indentação e um indicador de raiz, esquerda ou direita.
     * @param representacao StringBuilder que contém a representação da árvore.
     * @param nivel Profundidade do nó que será representada como indentação. 0 para raiz.
     */
    private void toStringVerticalRecursivo(StringBuilder representacao, int nivel, No<T> no, char posicao) {

        if (no.isVazio()) { return; }

        representacao.append(
                "%s%c %s%n"
                        .formatted(" ".repeat(nivel), posicao, no.getValor().toString())
        );

        toStringVerticalRecursivo(representacao, nivel + 1, no.getEsquerda(), 'E');
        toStringVerticalRecursivo(representacao, nivel + 1, no.getDireita(), 'D');

    }

    /**
     * Representação da árvore em texto na horizontal.
     * Nós irmãos ficam lado a lado.
     */
    public String toString() {
        if (raiz.isVazio()) { return ""; }

        // Para cada linha da representação é feita uma lista de nós do nível correspondente da árvore.
        ArrayList<ArrayList<No<T>>> linhas = new ArrayList<>(
                List.of(
                        new ArrayList<>(List.of(raiz))
                )
        );

        // A ‘String’ que será retornada.
        StringBuilder representacao = new StringBuilder();
        // Número de caracteres necessários para representar o maior valor da árvore.
        int larguraMaxima =  raiz.getValor().toString().length();
        boolean continuar = true;

        while (continuar) {
            ArrayList<No<T>> proximaLinha =  new ArrayList<>();
            continuar = false;

            for (No<T> no: linhas.get(linhas.size() - 1)) {
                if (no == null) {

                    proximaLinha.add(null);
                    proximaLinha.add(null);

                } else {

                    // Largura deste nó.
                    int larguraNo = no.getValor().toString().length();
                    if (larguraNo > larguraMaxima) {
                        larguraMaxima = larguraNo;
                    }

                    // Prepara a próxima linha.
                    if (no.getEsquerda().isVazio()) {
                        proximaLinha.add(null);
                    } else {
                        proximaLinha.add(no.getEsquerda());
                        continuar = true;
                    }
                    if (no.getDireita().isVazio()) {
                        proximaLinha.add(null);
                    } else {
                        proximaLinha.add(no.getDireita());
                        continuar = true;
                    }

                }
            }

            if (continuar) {
                linhas.add(proximaLinha);
            }
        }

        int numNosUltimaLinha = linhas.get(linhas.size() - 1).size();

        // Largura da última linha em função do espaço de um nó.
        // Espaçamento entre nós na última linha é um nó vazio.
        int larguraArvore =  (2 * numNosUltimaLinha) - 1;

        String espacoVazio = " ".repeat(larguraMaxima);
        String noVazio = "-".repeat(larguraMaxima);
        int profundidade = linhas.size();

        for (int i = 0; i < profundidade; i++) {
            ArrayList<No<T>> linha =  linhas.get(i);

            // Espaço entre nós em função do espaço de um nó. Ex.: 1 nó vazio, 2 nós vazios, ...
            int espacamento =  (int) Math.pow(2, (profundidade - i)) - 1;

            // Espaço que é adicionado à esquerda para centralizar o nó nesta linha.
            int larguraDestaLinha =
                    larguraArvore - // Largura da árvore em nós.
                            linha.size() - // Número de nós desta linha.
                            ((linha.size() - 1) * espacamento); // Espaço ocupado nesta linha entre os nós em função do espaço de um nó.

            // Adiciona a indentação antes da linha.
            representacao.append(
                   espacoVazio.repeat(larguraDestaLinha / 2)
            );

            for (int j = 0; j < linha.size(); j++) {
                No<T> no = linha.get(j);
                if (no == null) {
                    representacao.append(noVazio);
                } else {
                    representacao.append(String.format("%" + larguraMaxima + "s", no.getValor()));
                }
                if (j != linha.size() - 1) {
                    representacao.append(espacoVazio.repeat(espacamento));
                }
            }
            representacao.append("\n");
        }
        return representacao.toString();
    }

}
